internal/emigration/emigrate-nobles.lua

```diff
@@ -0,0 +1,352 @@
+
+    return siteGov
+end
+
+---@param unit df.unit
+---@param histFig df.historical_figure
+---@param oldSite df.historical_entity
+local function removeUnitFromSiteEntity(unit, histFig, oldSite)
+    local histFigId = histFig.id
+
+    -- free owned rooms
+    for i = #unit.owned_buildings-1, 0, -1 do
+        local tmp = df.building.find(unit.owned_buildings[i].id)
+        dfhack.buildings.setOwner(tmp, nil)
+    end
+
+    -- remove from workshop profiles
+    for _, bld in ipairs(df.global.world.buildings.other.WORKSHOP_ANY) do
+        for k, v in ipairs(bld.profile.permitted_workers) do
+            if v == unit.id then
+                bld.profile.permitted_workers:erase(k)
+                break
+            end
+        end
+    end
+    for _, bld in ipairs(df.global.world.buildings.other.FURNACE_ANY) do
+        for k, v in ipairs(bld.profile.permitted_workers) do
+            if v == unit.id then
+                bld.profile.permitted_workers:erase(k)
+                break
+            end
+        end
+    end
+
+    -- disassociate from work details
+    for _, detail in ipairs(df.global.plotinfo.labor_info.work_details) do
+        for k, v in ipairs(detail.assigned_units) do
+            if v == unit.id then
+                detail.assigned_units:erase(k)
+                break
+            end
+        end
+    end
+
+    -- unburrow
+    for _, burrow in ipairs(df.global.plotinfo.burrows.list) do
+        dfhack.burrows.setAssignedUnit(burrow, unit, false)
+    end
+
+    -- erase the unit from the fortress entity
+    for k,v in ipairs(oldSite.histfig_ids) do
+        if v == histFigId then
+            df.global.plotinfo.main.fortress_entity.histfig_ids:erase(k)
+            break
+        end
+    end
+    for k,v in ipairs(oldSite.hist_figures) do
+        if v.id == histFigId then
+            df.global.plotinfo.main.fortress_entity.hist_figures:erase(k)
+            break
+        end
+    end
+    for k,v in ipairs(oldSite.nemesis) do
+        if v.figure.id == histFigId then
+            df.global.plotinfo.main.fortress_entity.nemesis:erase(k)
+            df.global.plotinfo.main.fortress_entity.nemesis_ids:erase(k)
+            break
+        end
+    end
+
+    -- remove the old entity link and create new one to indicate former membership
+    histFig.entity_links:insert("#", {new = df.histfig_entity_link_former_memberst, entity_id = oldSite.id, link_strength = 100})
+    for k,v in ipairs(histFig.entity_links) do
+        if v._type == df.histfig_entity_link_memberst and v.entity_id == oldSite.id then
+            histFig.entity_links:erase(k)
+            break
+        end
+    end
+end
+
+-- adapted from emigration::desert()
+---@param unit df.unit
+---@param toSite df.world_site
+local function emigrate(unit, toSite)
+    local histFig = df.historical_figure.find(unit.hist_figure_id)
+    if histFig == nil then qerror("could not find histfig!") end
```

Comment:
this should be a recoverable error -- skip the unit if they (somehow) don't have a histfig

Reply:
Should this really be recovered from though? The assumption at this point is that `emigration()` is called on a noble, if a noble does not have an associated histfig wouldn't that indicate something wrong with the game's state?

Reply:
yeah, but that's not the player's problem. If they want to emigrate all nobles, and one is in an invalid state, we should still successfully emigrate the remaining nobles.

Reply:
Ah yeah, good point.